contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpSpanCustomizer.java

```diff
@@ -0,0 +1,101 @@
+ * limitations under the License.
+ */
+
+package io.opencensus.contrib.http;
+
+import io.opencensus.contrib.http.util.HttpTraceUtil;
+import io.opencensus.trace.Span;
+import io.opencensus.trace.SpanBuilder;
+import javax.annotation.Nullable;
+
+/**
+ * A helper class to customize an HTTP span.
+ *
+ * <p>It contains basic instrumentation implementation. See <a
+ * href="https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/HTTP.md">HTTP
+ * trace spec</a> for more information.
+ *
+ * <p>Users can extend this helper to customize builder and end status.
+ *
+ * @param <Q> the HTTP request entity.
+ * @param <P> the HTTP response entity.
+ * @since 0.13
+ */
+public class HttpSpanCustomizer<Q, P> {
+
+  /**
+   * Returns customized span name according to the request.
+   *
+   * <p>Note that in OpenCensus, {@link Span} is immutable and cannot be renamed.
+   *
+   * @param request the HTTP request entity.
+   * @param extractor the framework specific extractor to get information from the request.
+   * @return a string that represents the name of the span.
+   */
+  public String getSpanName(Q request, HttpExtractor<Q, P> extractor) {
+    // default span name
+    return "/" + extractor.getPath(request);
+  }
+
+  /**
+   * Customize the builder for the {@link Span}.
+   *
+   * <p>This allows user to set {@code Sampler}, parent {@code Span}, and the {@code recordEvents}
+   * decision.
+   *
+   * @param request the HTTP request entity.
+   * @param spanBuilder the {@link SpanBuilder} used to start the HTTP span.
+   * @param extractor the framework specific extractor to get information from the request.
+   * @since 0.13
+   */
+  public void customizeSpanBuilder(
+      Q request, SpanBuilder spanBuilder, HttpExtractor<Q, P> extractor) {
+    // do nothing by default.
+  }
+
+  /**
+   * Customize the span after it is started, but before sending (for client) or receiving (for
+   * server) the message.
+   *
+   * <p>This method will do nothing by default.
+   *
+   * @param request the HTTP request entity.
+   * @param span the span to be customized.
+   * @param extractor the framework specific extractor to get information from the request.
+   * @since 0.13
+   */
+  public void customizeSpanStart(Q request, Span span, HttpExtractor<Q, P> extractor) {
+    // do nothing by default.
+  }
+
+  /**
+   * Customize the span after receiving (for client) or sending (for server) the message.
+   *
+   * <p>This method will set the status by default. Users can override this to implement their own
+   * parsing logic.
+   *
+   * @param response the HTTP response entity. {@code null} means invalid response.
+   * @param error the error occurs when processing the response.
+   * @param span the span to be customized.
+   * @param extractor the framework specific extractor to get information from the request.
+   * @since 0.13
+   */
+  public void customizeSpanEnd(
```

Comment:
We now have a spec for what attributes we should record:
https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/HTTP.md

Reply:
My understanding of the spec is that it is just "recommended" but instrumenters (probably not us) still can decide what attributes they want to add.
If that is the case I would like to make the customizer `abstract` and implement a `OpenCensusHttpSpanCustomizer` to add those routines defined in our specs.

Reply:
In opencensus-scala we decided to have functions to set the basic http attributes as well since this is what you probably always want to set.

https://github.com/Sebruck/opencensus-scala/blob/master/http/src/main/scala/com/github/sebruck/opencensus/http/HttpAttributes.scala

Reply:
Updated. Now it contains basic routine. "Route" is not added because it is usually framework specific.